include/deal.II/matrix_free/hanging_nodes_internal.h

```diff
@@ -233,319 +227,408 @@ namespace internal
-                          else if (cell->line(local_line)->vertex_index(1) ==
-                                   neighbor_cell->line(local_line_neighbor)
-                                     ->vertex_index(0))
-                            {
-                              // We have a flipped edge of second type
-                              flipped = true;
-                            }
-                          else if (cell->line(local_line)->vertex_index(0) ==
-                                   neighbor_cell->line(local_line_neighbor)
-                                     ->vertex_index(1))
-                            {
-                              // We have a flipped edge of first type
-                              mask |= line_to_edge[local_line][3];
-                              flipped = true;
-                            }
-                          else
-                            ExcInternalError();
-
-                          // Copy the unconstrained values
-                          neighbor_dofs.resize(n_dofs_1d * n_dofs_1d *
-                                               n_dofs_1d);
-                          neighbor_cell->get_dof_indices(neighbor_dofs);
-                          // If the vector is distributed, we need to transform
-                          // the global indices to local ones.
-                          if (partitioner)
-                            for (auto &index : neighbor_dofs)
-                              index = partitioner->global_to_local(index);
-
-                          for (unsigned int i = 0; i < n_dofs_1d; ++i)
-                            {
-                              // Get local dof index along line
-                              const unsigned int idx =
-                                line_dof_idx(local_line, i, n_dofs_1d);
-                              dof_indices[idx] = neighbor_dofs
-                                [lexicographic_mapping[line_dof_idx(
-                                  local_line_neighbor,
-                                  flipped ? fe_degree - i : i,
-                                  n_dofs_1d)]];
-                            }
-
-                          // Stop looping over edge neighbors
-                          break;
-                        }
-                    }
-                }
-            }
-        }
+      bool cell_has_hanging_node_constraints = false;
+
+      const auto &fe = cell->get_fe();
+
+      std::vector<std::vector<unsigned int>>
+        component_to_system_index_face_array(fe.n_components());
+
+      for (unsigned int i = 0; i < fe.n_dofs_per_face(0); ++i)
+        component_to_system_index_face_array
+          [fe.face_system_to_component_index(i, /*face_no=*/0).first]
+            .push_back(i);
+
+      const auto component_to_system_index_face = [&](const auto comp,
+                                                      const auto i) {
+        return component_to_system_index_face_array[comp][i];
+      };
```

Comment:
Is using `()` rather than `[][]` worth the expense of another indirection and another function the reader needs to keep present?